testsuite/cluster-test/src/health/mod.rs

```diff
@@ -93,12 +96,24 @@ impl HealthCheckRunner {
         )
     }
 
-    /// Returns list of failed validators
-    pub fn run(&mut self, events: &[ValidatorEvent]) -> Vec<String> {
+    /// Takes a list of affected_validators. If there are validators which failed
+    /// which were not part of the experiment, then it returns an Err with a string
+    /// of all the unexpected failures.
+    /// Otherwise, it returns a list of ALL the failed validators
+    /// It also takes a bool parameter: only_print_on_failure. If this is set
+    /// to true, messages are printed only when there are failures.
+    /// If this is set to false, messages are always printed
+    pub fn run(
+        &mut self,
+        events: &[ValidatorEvent],
+        affected_validators_set: &HashSet<String>,
+        only_print_on_failure: bool,
+    ) -> failure::Result<Vec<String>> {
```

Comment:
@andll In the issue, you suggested a return type of `failure::Result<()>`, but there are clients which use the list of failed nodes. Example: `cluster-test/src/main.rs:497`. So I used return type of `failure::Result<Vec<String>>`

Reply:
`cluster-test/src/main.rs:497` seem to be something else now, can you give code snippet?

Reply:
Here the return value from HealthCheckRunner::run() is being use to break from the loop. This is in `fn run_single_experiment` in `cluster-test/src/main.rs`

```
            let unhealthy_validators;
            match self
                .health_check_runner
                .run(&events, &affected_validators, true)
            {
                Err(s) => bail!("Validators which were not under experiment failed : {}", s),
                Ok(r) => unhealthy_validators = r,
            }
            if unhealthy_validators.is_empty() {
                break;
            }
```

Reply:
Looks like it uses `unhealthy_validators` to generate error message, so we should be able to just reuse error message right?

Reply:
It is using the result (`unhealthy_validators`) to determine whether to break from the loop, not just for the error message.

Reply:
But criteria on whether to break the loop is same as criteria inside health check on whether to print error?
So we should not duplicate it...